code/modules/projectiles/gun.dm

```diff
@@ -2025,3 +2026,14 @@ not all weapons use normal magazines etc. load_into_chamber() itself is designed
 /obj/item/weapon/gun/proc/get_target()
 	RETURN_TYPE(/atom)
 	return target
+
+/obj/item/weapon/gun/ai_can_use(mob/living/carbon/human/user, datum/human_ai_brain/ai_brain)
+	if(istype(current_mag, /obj/item/ammo_magazine/internal) && (current_mag.current_rounds <= 0) && !ai_brain.weapon_ammo_search(ai_brain.primary_weapon))
+		return FALSE
+	else if(!current_mag && !ai_brain.weapon_ammo_search(ai_brain.primary_weapon))
```

Comment:
We can still have mag inside our weapon, when having no current ammo, which will result in (e.g.) not changing primary from smartgun (has no automatic mag ejection) to secondary
```suggestion
	else if(!(current_mag?.current_rounds <= 0) && !ai_brain.weapon_ammo_search(ai_brain.primary_weapon))
```